Webpanel remote MCDU

legoboyvdlp opened the pull request.

inuyaksa committed:
some logic optiomizing

inuyaksa committed:
Revert "some logic optiomizing"

This reverts commit b4e59e15b85c282cbecda6285cad9798009359db.

inuyaksa committed:
rework fix

inuyaksa committed:
MCDU web moved as webpanel

- MCDU as webpanel (not visible on PHI)
- New menu to access MCDU web
- Open via browser or QRCode
- Fixed MCDU web text selection #230

inuyaksa committed:
Merge branch 'dev' into rework

legoboyvdlp commented on Models/Instruments/MCDU/MCDU.nas:
```diff
@@ -410,9 +410,14 @@ var canvas_MCDU_base = {
 	},
 	updateCommon: func(i) {
 		page = pageProp[i].getValue();
+		var phase = fmgc.FMGCInternal.phase;
```
Comment on this: it is better instead of this, to do
},
**phaseTemp: 0,**
updateCommon: func(i) {
page = pageProp[i].getValue();
**me.phaseTemp** = fmgc.FMGCInternal.phase;



That improves on garbage collection since otherwise you have to create a new variable, phase, every time the func updateCommon() is run

legoboyvdlp replied:
And besides, accessing fmgc.FMGCInternal.phase is very fast + easy (well, maybe accessing me.phasetemp is even faster). It's not a big issue, compared to the use of getprop() or even getValue() to fetch a property node :-)

legoboyvdlp commented on Nasal/FMGC/FMGC.nas:
```diff
@@ -961,10 +1028,15 @@ var masterFMGC = maketimer(0.2, func {
 
 var reset_FMGC = func {
 	FMGCInternal.phase = 0;
+	FMGCNodes.phase.serValue(0);
```
typo. Should be setValue

legoboyvdlp commented on Nasal/FMGC/FMGC.nas:
```diff
@@ -961,10 +1029,15 @@ var masterFMGC = maketimer(0.2, func {
+
+};
+
+setlistener(FMGCNodes.phase, updateAirportRadios);
```
Nice idea. But perhaps, also do **,0,0);** at the end of the setlistener command so it doesn't accidentally get called all the time (it shouldn't, but just in case)

inuyaksa replied:
I agree, 0,0 is better :)

merspieler commented on gui/dialogs/mcduweb.xml:
```diff
@@ -0,0 +1,137 @@
+					var root = mcdu_canvas_dlg.createGroup();
+					mcdu_canvas_dlg.setColorBackground(1, 1, 1, 1);
+					var _url = "";
+					if( getprop("sim/http/running",0) == 1 ) _url = "http://localhost:" ~ getprop("/sim/http/options/listening-port") ~ "/aircraft-dir/WebPanel/mcdu.html";
```
This won't work!
You're referring to localhost but ask the user to to scan it with a phone/tablet-> different device.
You'd need to use either the IP address or the host name for this to work.

inuyaksa replied:
Only work if you are on the same network, your smartphone/tablet it suppose to use your home wifi connection.
It needs to clarify on doc.
It's possible to expose on external network, it's a bit complex but it's out of flightgear context.

inuyaksa replied:
After meditation... you are right. I need to detect local IP to let load mcdu on external devices.

inuyaksa replied:
@merspieler @legoboyvdlp I have update remote mcdu with input for local-ip request and some minor fixes.

merspieler commented on gui/dialogs/mcduweb.xml:
```diff
@@ -0,0 +1,137 @@
+					mcdu_canvas_dlg.setColorBackground(1, 1, 1, 1);
+					var _url = "";
+					if( getprop("sim/http/running",0) == 1 ) _url = "http://localhost:" ~ getprop("/sim/http/options/listening-port") ~ "/aircraft-dir/WebPanel/mcdu.html";
+					root.createChild("image").set("src", "https://qr.eletto.dev/" ~ _url).setSize(300,300).setTranslation(10,10); 					
```
I don't think, we should query such a service without the users consent (only generate that qr code if the user clicks a button like `Show QR code`

inuyaksa replied:
Needed because port it's user defined.
Do you think it is a concern to use external service?
Maps and weathers work with external service too.

merspieler replied:
Yes, cause it contains sensetiv information such as open ports and IP addresses...
Open ports are a potential security risk... so you shouldn't "announce" them to some service on the internet (except you know what you're doing -> not the case for normal users)

NewoIsTaken replied:
I think most people will only need it on local network. If they need it outside, they should be able to figure it out themselves. I can't think of a use case for having it outside your LAN.

merspieler replied:
True, but it doesn't mean, you should announce those things to the public... 
Remember stuff like the wannacry stuff.... it was only so successful cause people had their damn ports open.

inuyaksa replied:
@merspieler not our situation. The port opens it's the same for phi.
As you request for a button, I added button to generate qrcode on last version,

merspieler replied:
thanks

merspieler commented on gui/dialogs/mcduweb.xml:
```diff
@@ -0,0 +1,137 @@
+<!-- Copyright (c) 2020 Josh Davidson (Octal450) -->
+
+<!-- A3XX MCDU web launcher Dialog -->
+
```
I'm also missing a comment, that the firewall has to be configured accordingly

merspieler reviewed (changes requested):
QR Code needs a fix.

I didn't do a check ride or completely check the nasal code yet.

inuyaksa committed:
autoland light, phase as a nodes optimzed fmgc phase loop

inuyaksa committed:
property /instrumentation/radar-altimeter-difference-ft

Difference from 2 radar altimeter (corrected) in ft

inuyaksa committed:
Merge branch 'dev' into rework

legoboyvdlp commented on Nasal/FMGC/FMGC.nas:
```diff
@@ -603,88 +605,144 @@ var radios = maketimer(1, func() {
-	tempOverspeed = systems.ADIRS.overspeedVFE.getValue();
-	if (tempOverspeed != 1024) {
-		FMGCInternal.maxspeed = tempOverspeed - 4;
+	if (getprop("/systems/navigation/adr/computation/overspeed-vfe-spd") != 1024) {
```
please don't revert my changes :)  should use systems.ADIRS.overspeedVFE.getValue();

inuyaksa replied:
Sorry, I dunno I lost your changes. There is some other changes missing?

inuyaksa committed:
Revert "Merge branch 'dev' into rework"

This reverts commit 0140738582d99ff4f9e9c239106c98eea1024dc1, reversing
changes made to 23369f854b9cc1c703e94fcb7aee5ae0006a0d2a.

inuyaksa committed:
MCDU web, qrcode with local-it, fixes on html

- For mobile devices, you have to insert your local ip to generate a qrcode
- Fixed HTML for zoom action and double tap

inuyaksa committed:
typo

inuyaksa committed:
lost changes

inuyaksa committed:
A better autoland light - first attempt

merspieler reviewed (commented):
Looks better now...

2 Things tho:
* by removing the length check, we could allow IP or hostname to be used. The later is favorable as the IP might change so people would have to rescan the QR code even if they bookmarked the page on their device. Using the hostname would avoid that.
* In landscape mode (on a computer screen) the site becomes unusable as you have to scroll up and down constantly
